### DISCUSSION TITLE
Error after install INVO: Dispatcher has detected a cyclic routing

### DISCUSSION THREAD

**Discussion by Edd-G** [phalcon/cphalcon#16267] [Q&A]
Hi!

I install latest INVO (2.0.0).
On Linux &
Phalcon 5.1.2
Nginx/1.18.0
PHP-FPM 8.1

Nginx config get from Phalcon webserver setup:
https://docs.phalcon.io/5.0/en/webserver-setup#install-nginx

But any POST action (login form / comment form) getting error:
```
Dispatcher has detected a cyclic routing causing stability problems
#0 [internal function]: Phalcon\Mvc\Dispatcher->throwDispatchException()
#1 [internal function]: Phalcon\Dispatcher\AbstractDispatcher->dispatch()
#2 /home/edward/web/invo/public/index.php(46): Phalcon\Mvc\Application->handle()
#3 {main}
```
at this line
```
(new Application($di))
     ->handle($_SERVER['REQUEST_URI'])
     ->send()
;
```
I can't figure out what the problem is.

**Comment by s-ohnishi** [comment]
When I tried the INVO application, I encountered various problems when placing it under a subdirectory of the document root.
However, I don't think there was "cyclic routing" among them.
(The latest INVO has a lot of errors in v5, so I'm remodeling it. I'd like to submit a pull request when I have a prospect.)

I will try it again, so could you tell me the contents of the nginx configuration (*.conf) that you "actually" set and where you placed the INVO application?

**Comment by Edd-G** [comment]
Hi, [s-ohnishi](https://github.com/s-ohnishi)!
Thanks for your answer.

I install INVO in root directory, not in subdirectory.
Install path: /home/edward/development/web/invo.loc/
Local domain: invo.loc

Nginx conf based on official docs:
```
server {
	listen 80;
	server_name invo.loc;
	set $root_path '/home/edward/development/web/invo.loc/public';
	access_log off;
	error_log /var/log/nginx/invo.loc.error_log crit;
	root $root_path;
	client_max_body_size 100M;
	fastcgi_read_timeout 1800;
	location / {
		try_files $uri $uri/ /index.php?_url=$uri&$args;
	}
	include /etc/nginx/std-phalcon-php8.1.conf;
}
```
Next part of include /etc/nginx/std-phalcon-php8.1.conf
```
location ~ [^/]\.php(/|$) {
	fastcgi_pass unix:/run/php/php8.1-fpm.sock;
	fastcgi_index /index.php;
	include /etc/nginx/fastcgi_params;
	fastcgi_split_path_info ^(.+?\.php)(/.*)$;
	if (!-f $document_root$fastcgi_script_name) {
	    return 404;
	}
	fastcgi_param PATH_INFO $fastcgi_path_info;
	fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
}
```
Invo .env
```
DB_ADAPTER=Mysql
DB_HOST=localhost
DB_USERNAME=root
DB_PASSWORD=
DB_DBNAME=invo
DB_CHARSET=utf8

VIEWS_DIR=themes/invo/
BASE_URI=/

APP_URL=http://invo.loc
```
All works, but if I try open log in/sign up form or submit contact/register forms I get an error:
```
Dispatcher has detected a cyclic routing causing stability problems
#0 [internal function]: Phalcon\Mvc\Dispatcher->throwDispatchException()
#1 [internal function]: Phalcon\Dispatcher\AbstractDispatcher->dispatch()
#2 /home/edward/development/web/invo.loc/public/index.php(46): Phalcon\Mvc\Application->handle()
#3 {main}
```
Login/signup form shown after comment lines 38-39 in SessionController.php
```
public function indexAction(): void
    {
        // $this->tag->setDefault('email', 'demo');
        // $this->tag->setDefault('password', 'phalcon');
    }
```
However, if I try to submit the form, I get the error again: Dispatcher has detected a cyclic routing causing stability problems.
I have not yet investigated the error in detail.

**Reply by s-ohnishi** [reply]
Thanks for the additional information.
But "cyclic routing" did not occur.
Have you ever set any routing rules?

Just in case, here are the steps I tried.
- Download the latest version of the INVO application (invo-master.zip)
- Extract and move everything to /var/www/invotest in my test environment
- Give write permission to all directories under /var/www/invotest/var
- Copy .env.example and change to .env, and rewrite only DB_HOST (to match my environment)
- Run composer install (do not update)
- Add invo.loc to hosts
- Add what you wrote to the nginx configuration file

additional .conf (partially modified)
```
## test
server {
	listen 80;
	server_name invo.loc;
	set $root_path '/var/www/invotest/public';
#	access_log off;
#	error_log /var/log/nginx/invo.loc.error_log crit;
	error_log /var/log/nginx/invo.loc.error_log debug;
	root $root_path;
	client_max_body_size 100M;
	fastcgi_read_timeout 1800;
	location / {
		try_files $uri $uri/ /index.php?_url=$uri&$args;
	}
#	include /etc/nginx/std-phalcon-php8.1.conf;
    location ~ [^/]\.php(/|$) {
#		fastcgi_pass unix:/run/php/php8.1-fpm.sock;
		fastcgi_pass phpfpm:9000;
		fastcgi_index /index.php;
		include /etc/nginx/fastcgi_params;
		fastcgi_split_path_info ^(.+?\.php)(/.*)$;
		if (!-f $document_root$fastcgi_script_name) {
		    return 404;
		}
		fastcgi_param PATH_INFO $fastcgi_path_info;
		fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
	}
}
```
.env
```DB_ADAPTER=Mysql
# Use invo-mysql for dockerized environment or change it to where your db host is
DB_HOST=db
DB_USERNAME=phalcon
DB_PASSWORD=secret
DB_DBNAME=phalcon_invo
DB_CHARSET=utf8

VIEWS_DIR=themes/invo/
BASE_URI=/

APP_URL=http://invo.loc
```
And when I go to http://invo.loc I get a 403 Forbidden.
This is a natural result because the index is not set.
When I access http://invo.loc/index.php the page is displayed without any problem.

When I click Log In/Sign Up, I get an Internal Error.
After modifying SessionController.php, it will be displayed without any problem.
(However, the line number you indicated is off by one.)
```
    public function indexAction(): void
    {
//        $this->tag->setDefault('email', 'demo');
//        $this->tag->setDefault('password', 'phalcon');
    }
```
![SessionController](https://user-images.githubusercontent.com/6782530/214509792-479e5b76-f770-4606-b12f-26952ad39295.png)
![session_index](https://user-images.githubusercontent.com/6782530/214509846-3b6d836f-7a3e-4504-b376-0892b75260e5.png)

**Reply by s-ohnishi** [reply]
Sorry, I forgot to mention the environment.
Nginx: 1.22.1
PHP: 8.1.14
Phalcon: 5.1.3
Please forgive me for not being able to match the environment you indicated.

**Comment by Edd-G** [comment]
No, I not set any routing rules.

To see the error try the following:

1. Try to submit any form: login, register or contact form.
You will see: Internal Error
This error is displayed by the ErrorsController and gives little information.
2. Try comment all class in controller/ErrorsController.php

```
// class ErrorsController extends ControllerBase
// {
//     public function initialize()
//     {
//         $this->tag->title()
//                   ->set('Oops!')
//         ;

//         parent::initialize();
//     }

//     public function show404Action(): void
//     {
//         $this->response->setStatusCode(404);
//     }

//     public function show401Action(): void
//     {
//         $this->response->setStatusCode(401);
//     }

//     public function show500Action(): void
//     {
//         $this->response->setStatusCode(500);
//     }
// }
```
I have not yet figured out how to display more detailed information about the error.
So I just commented out the whole class)

Now submit any form and you can see:
Dispatcher has detected a cyclic routing causing stability problems

**Reply by s-ohnishi** [reply]
I got it.
This part is the reason.
>Try comment all class in controller/ErrorsController.php

Because the class for displaying errors was disabled without correcting the cause of the internal error, "cyclic routing" may have occurred by trying to call this class many times.

First, uncomment everything in ErrorsController.
And comment out setDefault() in SessionController::indexAction() which is the real cause.
```PHP
    public function indexAction(): void
    {
//        $this->tag->setDefault('email', 'demo');
//        $this->tag->setDefault('password', 'phalcon');
    }
```
(The controller's tag property has changed from Phalcon\Tag to Phalcon\Html\TagFactory and setDefault() does not exist)

I think that "cyclic routing" will not appear with this.
Also, setDefault() is used by other controllers, so they also need to be fixed.

The INVO application continues to be published as "latest" without supporting v5, and there are many bugs to use with v5.
For example:
- the expected argument is not passed
- the form method is specified as get, but the program is written as post
- etc...

**Reply by Edd-G** [reply]
Hi!
Thanks, you right!
1. I revert all back in ErrorsController.
2. Comment this IF block in plugins/NotFoundPlugin.php
```
// if ($dispatcher->getControllerName() !== 'errors') {
//     $dispatcher->forward([
//         'controller' => 'errors',
//         'action'     => 'show500',
//     ]);
// }
```
After that, enable debug in public/index.php
comment: try / catch
and add this
```
use Phalcon\Support\Debug;

(new Debug())->listen(true, false); // exceptions - bool, lowSeverity - bool
```
Now I can see all errors and I can fix them.

$this->tag->setDefault()
replace with
```
    public function indexAction(): void
    {
        Tag::setDefault('email', 'demo');
        Tag::setDefault('password', 'phalcon');
    }
```
and add
use Phalcon\Tag;

**Reply by s-ohnishi** [reply]
Glad you were able to fix the error.

I think that using Phalcon\Tag statically should only be used to avoid the problems you face.
Phalcon\Tag is no longer mainstream in v5, and it seems that it will move to Phalcon\Html\TagFactory in the future.
If you modify the sample application, I think it would be a good idea to actively use TagFactory to accumulate knowledge.
